### PR TITLE
BUG: Ensure same index is returned for slow and fast path in groupby.apply

### LINKED ISSUE #13056: API: inconsistent return format of groupby apply
related issues:
- [x] #13121 
- [x] #5839
- [x] #9867
- [x] #13255 
- [x] #20066 
- [x] #14927 (about detecting mutating functions)

_Thank you very much for provide us pandas which is really a good packages I have used!_
### There were several times when I use pandas but get inconsistent return format which finally leads to a break!
- for example:

``` python
import numpy as np
import pandas as pd
df = pd.DataFrame(np.random.randn(10,1), columns = ['x'])
df['y'] = 1
df.groupby('y').apply(lambda x: x.x)

#x        0         1         2         3         4         5         6  \
#y                                                                        
#1 -1.12114  0.679616 -1.392863  0.032637 -0.051134  0.594201 -0.238833   

#x        7        8         9  
#y                              
#1  0.95173  1.07469 -0.062198  

df2 = df.copy()
df2['y'] = np.random.randint(1,3,10)
# do something similar like above
df2.groupby('y').apply(lambda x: x.x)
#y   
#1  3    0.032637
#   8    1.074690
#2  0   -1.121140
#   1    0.679616
#   2   -1.392863
#   4   -0.051134
#   5    0.594201
#   6   -0.238833
#   7    0.951730
#   9   -0.062198
#Name: x, dtype: float64
```

even though, I knew the way to avoid this issue,  I still believe it is better to return the same format. It's really annoying when some inconsistent return format happened which force me to   look backwards to use some awkward methods to avoid it.

So, if possible, I do suggest the above function will only return one kind of format.

### LINKED ISSUE #14927: DataFrame.groupby.apply returns different results with copy lambda functions.
#### Code Sample, a copy-pastable example if possible

```python
In [32]: df = pd.DataFrame({"g":[1, 2, 2, 2], "a":[1, 2, 3, 4], "b":[5, 6, 7, 8]})
    ...: 

In [33]: df.groupby('g').apply(lambda x: x)
Out[33]: 
   a  b  g
0  1  5  1
1  2  6  2
2  3  7  2
3  4  8  2

In [34]: df.groupby('g').apply(lambda x: x[:])
Out[34]: 
     a  b  g
g           
1 0  1  5  1
2 1  2  6  2
  2  3  7  2
  3  4  8  2

In [35]: df.groupby('g').apply(lambda x: x.copy(deep=False))
Out[35]: 
   a  b  g
0  1  5  1
1  2  6  2
2  3  7  2
3  4  8  2

In [36]: df.groupby('g').apply(lambda x: x.copy(deep=True))
Out[36]: 
     a  b  g
g           
1 0  1  5  1
2 1  2  6  2
  2  3  7  2
  3  4  8  2

```
#### Problem description

xref #9867

Issue #9946 (In[36])  no longer raises a `ValueError` on master 0.19.1, but I am curious why these 4 cases return 2 different results. 

#### Expected Output

Not sure since I don't have a personal use-case for this operation, but I'd expect these operations to return the same result unless I am missing something with using a deepcopy.
 
#### Output of ``pd.show_versions()``

<details>
```

commit: 3ccb50131b698bccee21780c660a70bed87396d0
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-45-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.19.0+228.g3ccb501
nose: 1.3.7
pip: 8.1.2
setuptools: 27.2.0
Cython: 0.24.1
numpy: 1.11.2
scipy: None
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.4.8
patsy: None
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
pandas_datareader: None
```
</details>

### LINKED ISSUE #31612: Groupby apply result shape depends on internal apply path
#### Problem description

The shape returned by groupby.apply currently depends on the path the internal apply takes (fast apply vs slow apply) which is opaque to the user. The following examples show the behaviour for two simple examples where the fast path returns the group key as index while the slow path returns the original index.

#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({'A': [0, 0, 1], 'b': range(3)})

def slow(group):
       # slow apply because of check `result is input`, c.f. https://github.com/pandas-dev/pandas/blob/44782c0809e296a8e57b7f77d963e999c7e0f4a7/pandas/_libs/reduction.pyx#L494
       return group

def fast(group):
       return group.copy()

```
```
   ...: df.groupby("A").apply(slow)
   ...:
Out[3]:
   A  B
0  0  0
1  0  1
2  1  2

In [4]: df.groupby("B").apply(fast)
Out[4]:
     A  B
B
0 0  0  0
1 1  0  1
```

#### Expected Output

A transparent and consistent output data type. In particular, the behaviour should not be coupled to private, performance related internals.

#### Related issues

There are a lot of related issues which may or may not have the same root cause. Here an excerpt

#12824
#14538
#12977
#28662

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.1.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.0.0
numpy            : 1.16.5
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.1.0.post20200119
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.12.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by fjetter**
This _fixes_ the internal check to be consistent with the slow apply path

- [x] closes #31612
- [x] closes #14927
closes #13056 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_apply.py
Comment (MarcoGorelli) [inline pandas/tests/groupby/test_apply.py]: Thanks!

Can you add the issue number here? (GH 31612)

File: doc/source/whatsnew/v1.0.1.rst
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.1.rst]: GroupBy isn't in the top level. Usually we write this as \`\`.groupby().apply()\`\`.

And as a user, I wouldn't know what to make of "depending on the internals".

Comment (TomAugspurger) [reply doc/source/whatsnew/v1.0.1.rst]: Perhaps something like "where the result shape was incorrect when the output index was not identical to the input index"?

Comment (fjetter) [reply doc/source/whatsnew/v1.0.1.rst]: Changed it to :meth:`core.groupby.DataFrameGroupBy.apply` which should be rendered properly by sphinx

File: pandas/_libs/reduction.pyx
Comment (fjetter) [inline pandas/_libs/reduction.pyx]: Is this something we should perform benchmarking before merging?

File: doc/source/whatsnew/v1.0.1.rst
Comment (jreback) [inline doc/source/whatsnew/v1.0.1.rst]: can you put this under the groupby section

File: pandas/_libs/reduction.pyx
Comment (jreback) [reply pandas/_libs/reduction.pyx]: this is cheap

File: pandas/tests/groupby/test_apply.py
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: can you give this a real name

File: pandas/tests/groupby/test_apply.py
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: either parameterize this test, or call one of these expected and do 

result =
assert_frame_equal

result = ....

...
and so on

File: pandas/tests/groupby/test_apply.py
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: see if we have similar fast/slow tests; place these by them (if not ok).

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
are any of the other issues you referenced closed by this? (aside from the ones you listed)

**Comment by jreback** [comment]
is this a regression from 0.25.x?

**Comment by fjetter** [comment]
> are any of the other issues you referenced closed by this? (aside from the ones you listed)
The issues I referenced in the issue are already closed. I primarily referenced them to highlight this as a hotspot.

> is this a regression from 0.25.x
I believe this is not a regression and has been around for a while now. I'll check anyhow to confirm.

- [ ] check if it a regression

**Comment by jorisvandenbossche** [comment]
If this is not a regression, let's keep it for 1.1 then, I would say

**Comment by TomAugspurger** [comment]
Agreed.

File: doc/source/whatsnew/v1.0.1.rst
Comment (jreback) [reply doc/source/whatsnew/v1.0.1.rst]: can you do this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls merge master

**Comment by jbrockmendel** [comment]
I've found that if I disable fast_apply, then there is a test failure in tests.groupby.test_whitelist for `tshift`.  any idea why this doesnt affect that?

File: doc/source/whatsnew/v1.0.1.rst
Comment (jreback) [reply doc/source/whatsnew/v1.0.1.rst]: move to 1.1

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok, but have comments

File: doc/source/whatsnew/v1.0.1.rst
Comment (jreback) [reply doc/source/whatsnew/v1.0.1.rst]: can you do this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and ping on green and we can look

**Comment by fjetter** [comment]
sorry, was a bit swamped. Will rebase and address the comments

File: doc/source/whatsnew/v1.0.1.rst
Comment (fjetter) [reply doc/source/whatsnew/v1.0.1.rst:44]: done

File: pandas/tests/groupby/test_apply.py
Comment (fjetter) [reply pandas/tests/groupby/test_apply.py:875]: done

File: pandas/tests/groupby/test_apply.py
Comment (fjetter) [reply pandas/tests/groupby/test_apply.py:879]: done

File: pandas/tests/groupby/test_apply.py
Comment (fjetter) [reply pandas/tests/groupby/test_apply.py]: There are a few tests further up the file wich affect similar code paths, e.g. `test_fast_apply` or `test_group_apply_once_per_group` which seems a suitable place

**Comment by fjetter** [comment]
@jreback I rebased but it seems there is a (probably unrelated) test failure on the azure pipelines with `test_tab_complete_*` tests. Is this a known issue?

**Comment by jbrockmendel** [comment]
> there is a (probably unrelated) test failure on the azure pipelines with test_tab_complete_* tests. Is this a known issue?

It is unrelated, but not yet trouble-shot.

**Comment by simonjayhawkins** [comment]
> > there is a (probably unrelated) test failure on the azure pipelines with test_tab_complete_* tests. Is this a known issue?
> 
> It is unrelated, but not yet trouble-shot.

@fjetter #33567 has been fixed. can you merge upstream/master.

**Comment by fjetter** [comment]
I rebased on master and kept the changelog for 1.1.0

File: doc/source/whatsnew/v1.1.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:830]: can you make this more user informative, e.g. what case does this happen

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, if you can update the whatsnew note a bit (ok to make a bit longer; re-reading this is not that common i think), but still want to signal the intentions to the user a bit more.

File: doc/source/whatsnew/v1.1.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:830]: can you also list the other closed issue here #14927

**Comment by jreback** [comment]
ping on green.

**Comment by fjetter** [comment]
Trying to rewrite the changelog. pretty difficult to actually list all tickets affected by this since most users don't go into depths that far.
There is also an umbrella issue which tracks these kind of things which we may be able to close now #13056 another opinion appreciated)

**Comment by jreback** [comment]
> Trying to rewrite the changelog. pretty difficult to actually list all tickets affected by this since most users don't go into depths that far.
> There is also an umbrella issue which tracks these kind of things which we may be able to close now #13056 another opinion appreciated)

what you wrote is ok, the idea is really to just list the affected tickets so that users can click thru if desired. when we close an issue we want it referenced in the whatsnew and a test to confirm :->

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @fjetter very nice!

**Pull request merged**
